pkg/agent/controller/networkpolicy/priority.go

```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+	costSiftDown, costSiftUp, emptiedSlotsLow, emptiedSlotsHigh := 0, 0, 0, 0
+	for reassignBoundLow >= PolicyBottomPriority && emptiedSlotsLow < target {
+		if _, exists := pa.ofPriorityMap[reassignBoundLow]; exists {
+			costSiftDown++
+		} else {
+			emptiedSlotsLow++
+			costMap[emptiedSlotsLow] = &reassignCost{reassignBoundLow, upperBound - 1, costSiftDown}
+		}
+		reassignBoundLow--
 	}
-	return nil, nil
-}
-
-// getNextVacantOFPriority returns the first higher ofPriority that is currently vacant in the table,
-// starting from the input ofPriority.
-// Note that if the input ofPriority itself is vacant, it will simply return that ofPriority.
-// The search is incrementally against all ofPriorities available as it assumes the table is sparse in most cases.
-func (pa *priorityAssigner) getNextVacantOFPriority(ofPriority uint16) *uint16 {
-	for i := ofPriority; i <= PolicyTopPriority; i++ {
-		// input ofPriority will be greater than or equal to PolicyBottomPriority
-		if _, exists := pa.ofPriorityMap[i]; !exists {
-			return &i
+	for reassignBoundHigh <= PolicyTopPriority && emptiedSlotsHigh < target {
+		if _, exists := pa.ofPriorityMap[reassignBoundHigh]; exists {
+			costSiftUp++
+		} else {
+			emptiedSlotsHigh++
+			// visit costMap in the reverse direction
+			mapIndex := target - emptiedSlotsHigh
+			c, ok := costMap[mapIndex]
+			// only add to the costMap if the counterpart cost is available. i.e. if the target is 4, and cost for
+			// using 2 empty slots high is computed, it does not make sense to store this cost if there's no entry
+			// for cost that uses 2 empty slots low (indicating no 2 empty slots can be found starting from lowerBound).
+			if ok {
+				c.cost = costSiftDown + costSiftUp
+				c.upperBound = reassignBoundHigh
+			} else if mapIndex == 0 {
+				costMap[mapIndex] = &reassignCost{lowerBound + 1, reassignBoundHigh, costSiftUp}
+			}
 		}
+		reassignBoundHigh++
 	}
-	return nil
-}
-
-// getLastOccupiedOFPriority returns the first ofPriority lower than the input ofPriority that is
-// currently occupied in the table, as well as the corresponded Priority.
-// Note that the function must return a ofPriority that is lower than the input ofPriority.
-// The search is based on sortedOFPriorities as it assumes the table is sparse in most cases.
-func (pa *priorityAssigner) getLastOccupiedOFPriority(ofPriority uint16) (*uint16, *types.Priority) {
-	idx := sort.Search(len(pa.sortedOFPriorities), func(i int) bool { return ofPriority <= pa.sortedOFPriorities[i] })
-	if idx > 0 {
-		lastOccupied := pa.sortedOFPriorities[idx-1]
-		priority := pa.ofPriorityMap[lastOccupied]
-		return &lastOccupied, &priority
+	minCost, minCostIndex := math.MaxInt32, 0
+	for i := target; i >= 0; i-- {
+		if cost, exists := costMap[i]; exists && cost.cost < minCost {
+			// make sure that the reassign range makes up for all Priorities to be registered.
+			if int(cost.upperBound-cost.lowerBound)+1 == numNewPriorities+cost.cost {
```

Comment:
isn't this guaranteed by your algorithm above? If yes, wouldn't it make sense to log something or return an error if this doesn't hold?

Reply:
No it is not guaranteed by the algorithm above and there should not be an error thrown. Consider a case where we need to find 3 empty spots below lowerBound and/or above upperBound. In the first pass, the algorithm tries to fill the table on where the lowerBound would be at if we were to use 1, 2 and 3 lower empty spots. For 1 and 2, when a cost is found, the upperBound will be temporarily set to the current upperBound which has not shifted at all. Those costs however only makes sense if in the second pass, the algorithm is able to find cost of getting 2 and 1 empty spots from above, in which case the cost and new upperBound will be updated. 
In other words, when `int(cost.upperBound-cost.lowerBound)+1 != numNewPriorities+cost.cost`, it simply means that this combination of (use x empty slots below, use y empty slots above) is not feasible based on the current map, but it's not an error.